src/actix/certificate_helpers.rs

```diff
@@ -1,126 +1,172 @@
+        return Err(io::Error::new(io::ErrorKind::Other, "No private key found"))
+    };
+    let private_key = rustls::PrivateKey(pkey);
+    let signing_key = rustls::sign::any_supported_type(&private_key)
+        .map_err(|err| io::Error::new(io::ErrorKind::Other, err))?;
 
+    // Construct certified key
+    let certified_key = CertifiedKey::new(certs, signing_key);
+    Ok(Arc::new(certified_key))
+}
+
+/// Generate an actix server configuration with TLS
+///
+/// Uses TLS settings as configured in configuration by user.
+pub fn actix_tls_server_config(settings: &Settings) -> io::Result<ServerConfig> {
+    let config = ServerConfig::builder().with_safe_defaults();
     let tls_config = settings
         .tls
         .clone()
         .ok_or_else(Settings::tls_config_is_undefined_error)?;
 
-    let verify_client_cert = settings.service.verify_https_client_certificate;
-    let ssl_context_holder = RwLock::new(SslContextHolder::new(
-        &tls_config,
-        verify_client_cert,
-        Duration::from_secs(tls_config.cert_ttl),
-    )?);
-
-    // Use set_servername_callback to implement dynamic certificate loading and refresh
-    acceptor.set_servername_callback(move |ssl, _alert| -> Result<(), SniError> {
-        {
-            let reader = ssl_context_holder.read();
-            if let Some(ssl_context) = reader.try_get_ssl_context() {
-                ssl.set_ssl_context(ssl_context).map_err(|error_stack| {
-                    log::error!("Failed to set SSL context: {}", error_stack);
-                    SniError::ALERT_FATAL
-                })?;
-                return Ok(());
-            }
-        }
+    // Verify client CA or not
+    let config = if settings.service.verify_https_client_certificate {
+        let mut root_cert_store = RootCertStore::empty();
+        let ca_certs: Vec<Vec<u8>> = with_buf_read(&tls_config.ca_cert, rustls_pemfile::certs)?;
+        root_cert_store.add_parsable_certificates(&ca_certs[..]);
+        config.with_client_cert_verifier(AllowAnyAuthenticatedClient::new(root_cert_store))
+    } else {
+        config.with_no_client_auth()
+    };
 
-        // If getting the SSL context failed, try to get it again with refreshing
-        let mut writer = ssl_context_holder.write();
-        let ssl_context = writer
-            .get_ssl_context_or_refresh()
-            .map_err(|_refresh_error| {
-                log::error!("Failed to refresh certificates!");
-                SniError::ALERT_FATAL
-            })?;
-        ssl.set_ssl_context(ssl_context).map_err(|error_stack| {
-            log::error!("Failed to set SSL context: {}", error_stack);
-            SniError::ALERT_FATAL
-        })?;
-        Ok(())
-    });
-    Ok(acceptor)
+    // Configure rotating certificate resolver
+    let ttl = match tls_config.cert_ttl {
+        None | Some(0) => None,
+        Some(seconds) => Some(Duration::from_secs(seconds)),
+    };
```

Comment:
It might make sense to have log INFO here to tell the user whether certificate rotation is enabled (with which TTL).

It seems to be an important point that should easily visible from the logs.
WDYT?

Reply:
In a chat we agreed that it would be good to show a single line on startup, telling the user whether TLS is enabled, including the TTL. That is a bit more general.

Reply:
Implemented inc6068db3e2f74a626e1c41d9ab649896ce36ca5b.